### PR TITLE
Switch to native-routing in GKE guide

### PR DESCRIPTION
**Pull request by tgraf**
This changes the GKE installation instructions to default to direct-routing and provides the required documentation around it.

### CODE REVIEW

**Comment by tgraf** [comment]
test-gke

File: test/k8sT/DatapathConfiguration.go [hunk 1]
```diff
@@ -304,6 +304,36 @@ var _ = Describe("K8sDatapathConfig", func() {
...
+			})
+
+			Expect(testPodConnectivityAcrossNodes(kubectl)).Should(BeTrue(), "Connectivity test between nodes failed")
+		})
+	})
+
+	Context("AutoDirectNodeRoutes", func() {
 		BeforeEach(func() {
 			SkipIfIntegration(helpers.CIIntegrationFlannel)
 			SkipIfIntegration(helpers.CIIntegrationGKE)
```
Comment (aanm) [inline test/k8sT/DatapathConfiguration.go:330]: Does this still needs to be skipped?

Comment (tgraf) [reply test/k8sT/DatapathConfiguration.go:330]: Yes, a single L2 network is not given. In fact, auto-direct-node routes is not required at all on GKE as the default route will point to a router that can route all PodIPs as announced in Node.Spec.PodCIDR[s] so we can simply disable tunneling to achieve native routing across everything.

**Comment by tgraf** [comment]
test-gke

File: Documentation/gettingstarted/k8s-install-gke.rst [hunk 2]
```diff
@@ -72,6 +73,13 @@ correct one (see `GKE RBAC documentation`_ for more info).
 Deploy Cilium
 =============
 
+Extract the Cluster CIDR to enable native-routing:
+
+.. code:: bash
+
+    NATIVE_CIDR=$(gcloud container clusters describe $CLUSTER_NAME | grep clusterIpv4Cidr | awk '{print $2}')
```
Comment (aanm) [inline Documentation/gettingstarted/k8s-install-gke.rst]: small nit: `grep clusterIpv4Cidr -i`

**Review by aanm — APPROVED** [review]

File: Documentation/gettingstarted/k8s-install-gke.rst [hunk 3]
```diff
@@ -88,7 +96,11 @@ below. This will ensure all pods are managed by Cilium.
       --set global.nodeinit.enabled=true \\
       --set nodeinit.reconfigureKubelet=true \\
       --set nodeinit.removeCbrBridge=true \\
-      --set global.cni.binPath=/home/kubernetes/bin
+      --set global.cni.binPath=/home/kubernetes/bin \\
+      --set global.k8s.requireIPv4PodCIDR=true \\
+      --set global.tunnel=disabled \\
+      --set global.endpointRoutes.enabled=true \\
+      --set global.native-routing-cidr=$NATIVE_CIDR
```
Comment (joestringer) [inline Documentation/gettingstarted/k8s-install-gke.rst:81]: Do we also need to do something explicit about disabling cilium-health here? I think we'll allocate an address to it that's not routable.

**Review by joestringer — CHANGES_REQUESTED** [review]
One minor comment below.

File: Documentation/gettingstarted/k8s-install-gke.rst (hunk 3 above)
Comment (tgraf) [reply Documentation/gettingstarted/k8s-install-gke.rst:81]: The great news is that we don't have to. All IPs in the PodCIDR handed to us are fully-routable. The connectivity health checking is working as expected:
```
KVStore:                Ok   Disabled
Kubernetes:             Ok   1.15+ (v1.15.11-gke.9) [linux/amd64]
Kubernetes APIs:        ["CustomResourceDefinition", "cilium/v2::CiliumClusterwideNetworkPolicy", "cilium/v2::CiliumEndpoint", "cilium/v2::CiliumNetworkPolicy", "cilium/v2::CiliumNode", "core/v1::Endpoint", "core/v1::Namespace", "core/v1::Pods", "core/v1::Service", "networking.k8s.io/v1::NetworkPolicy"]
KubeProxyReplacement:   Probe   [NodePort (HYBRID, 30000-32767, XDP: NONE), HostPort, ExternalIPs, HostReachableServices (TCP, UDP)]
Cilium:                 Ok      OK
NodeMonitor:            Disabled
Cilium health daemon:   Ok
IPAM:                   IPv4: 7/255 allocated from 10.68.2.0/24,
Controller Status:      36/36 healthy
Proxy Status:           OK, ip 10.68.2.184, 3 redirects active on ports 10000-20000
Hubble:                 Disabled
Cluster health:         3/3 reachable   (2020-04-21T20:52:18Z)
```

**Review by joestringer — APPROVED** [review]

**Comment by tgraf** [comment]
We'll need a different CI strategy for GKE. The individual tests can't overwrite the overrides yet that are put in place for GKE so we currently can't say `tunnel=disabled` by default if the Integration is GKE and then enable it for one particular test. Will check with @nebril

**Comment by tgraf** [comment]
test-gke

**Comment by tgraf** [comment]
test-gke

**Comment by tgraf** [comment]
test-gke

File: Documentation/concepts/datapath/gke.rst [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+Masquerading
+   All traffic not staying with the ``native-routing-cidr`` (defaults to the
+   Cluster CIDR) will be masqueraded to the node's IP address to become
+   publicly routable.
+
+Load-balancing
+   ClusterIP load-balancing will be performed using BPF for all version of GKE.
+   Starting with >= GKE v1.15 or when running a Linux kernel >= 4.19, all
+   NodePort/ExernalIP/HostPort will be performed using a BPF implementation as
```
Comment (tklauser) [inline Documentation/concepts/datapath/gke.rst]: Typo:
```suggestion
   NodePort/ExternalIP/HostPort will be performed using a BPF implementation as
```

File: Documentation/concepts/datapath/gke.rst [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  * ``tunnel: disabled``: Enable native routing mode
+  * ``enable-endpoint-routes: true``: Enable per-endpoint routing on the node
+  * ``blacklist-conflicting-routes: false``: Disable blacklisting of IPs
+    which collide with a local route
+  * ``enable-local-node-route: false``: Disable installation of the local node route
+
+* ``native-routing-cidr: x.x.x.x/y``: Set the CIDR in which native routing
+  is supported.
+
+See the getting started guide :ref:`k8s_install_gke` to install Cilium Google
```
Comment (tklauser) [inline Documentation/concepts/datapath/gke.rst]: Nit:

```suggestion
See the getting started guide :ref:`k8s_install_gke` to install Cilium on Google
```

File: Documentation/concepts/datapath/native_routing.rst [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+encapsulation.
+
+.. image:: native_routing.png
+    :align: center
+
+***************************
+Requirements on the network
+***************************
+
+* In order to run the native routing mode. The network connecting the hosts on
```
Comment (tklauser) [inline Documentation/concepts/datapath/native_routing.rst]: Nit: punctiuation

```suggestion
* In order to run the native routing mode, the network connecting the hosts on
```

File: Documentation/concepts/datapath/native_routing.rst [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+************
+Masquerading
+************
+
+Native routing is typically enabled in the context of a virtual network with
+private IP addresses. For any destination outside of the virtual network,
+traffic must typically be masqueraded. This is done by setting ``masquerade:
+true`` (default). In order to exclude the entire CIDR of the virtual network,
+the datapath must be told the CIDR within native routing is supported. This is
```
Comment (tklauser) [inline Documentation/concepts/datapath/native_routing.rst]:
```suggestion
the datapath must be told the CIDR within which native routing is supported. This is
```

File: Documentation/concepts/ipam/kubernetes.rst [hunk 8]
```diff
@@ -39,5 +45,21 @@ Annotation                          Description
...
+.. _hostscope_configuration:
+
+*************
+Configuration
+*************
+
+The following ConfigMap options exist to configure Kubernetes hostscope:
+
+ * ``ipam: kubernetes``: Enables Kubernetes IPAM mode. Enabling this option will
+   automatically enabled ``k8s-require-ipv4-pod-cidr`` if ``enable-ipv4`` is
```
Comment (tklauser) [inline Documentation/concepts/ipam/kubernetes.rst]:
```suggestion
   automatically enable ``k8s-require-ipv4-pod-cidr`` if ``enable-ipv4`` is
```

**Review by tklauser — COMMENTED** [review]
Some small spelling/punctuation/typo remarks in the docs.

**Comment by tgraf** [comment]
test-me-please

**Pull request merged**